pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -362,6 +362,49 @@ func (c *AntreaIPAMController) processNextStatefulSetWorkItem() bool {
 	return true
 }
 
+func (c *AntreaIPAMController) updateIPPoolCounters(obj interface{}) {
+	ipPool := obj.(*crdv1a2.IPPool)
+
+	allocator, err := poolallocator.NewIPPoolAllocator(ipPool.Name, c.crdClient, c.ipPoolLister)
+
+	if err != nil {
+		klog.Warningf("Failed to initialize allocator for IPPool %s", ipPool.Name)
+	}
+
+	// Total is fetched from allocator as here are trapped changes to CRD, e.g addition of new IPRange
+	total := allocator.Total()
+
+	// Used is gathered from IP allocation status within the CRD - as it can be set by each one of the agents
+	used := len(ipPool.Status.IPAddresses)
+
+	// Update the status within the CRD
+	pool, err := c.ipPoolLister.Get(ipPool.Name)
+	if err != nil {
+		klog.Warningf("Failed to retrieve pool %s from CRD", ipPool.Name)
+		return
+	}
+
+	// If update has no effect, exit
+	if pool.Status.Usage.Used == used && pool.Status.Usage.Total == total {
+		return
+	}
+	pool.Status.Usage.Used = used
+	pool.Status.Usage.Total = total
+
+	_, err = c.crdClient.CrdV1alpha2().IPPools().UpdateStatus(context.TODO(), pool, metav1.UpdateOptions{})
```

Comment:
1. event handlers are executed sequentially, doing all jobs here may not scale well, especially when there is an external API call. 
2. Every update to a pool will call the event handler once, doing all jobs here is not as effecient as doing it in workers, as multiple updates to a pool will only trigger one processing if the time the events happen is close.
A more typical and efficient way is:
Event handlers extract the pool's key and enqueue it; multiple workers get keys from the queue and do the actual job.